addons/website_sale_delivery/controllers/main.py

```diff
@@ -32,7 +33,10 @@ def update_eshop_carrier(self, **post):
     @http.route(['/shop/carrier_rate_shipment'], type='json', auth='public', methods=['POST'], website=True)
     def cart_carrier_rate_shipment(self, carrier_id, **kw):
         order = request.website.sale_get_order(force_create=True)
-        assert int(carrier_id) in order._get_delivery_methods().ids, "unallowed carrier"
+
+        if not int(carrier_id) in order._get_delivery_methods().ids:
+            raise UserError(_('It seems that a delivery method is not compatible with your address. Please refresh the page and try again.'))
```

Comment:
Doesn't that end up in a 500 Error on the website anyway?

Reply:
Nope you get the following traceback
<details>
	<summary>Toggle</summary>

> Error:
> Odoo Server Error
> 
> Traceback (most recent call last):
>   File "/data/build/odoo/odoo/http.py", line 624, in _handle_exception
>     return super(JsonRequest, self)._handle_exception(exception)
>   File "/data/build/odoo/odoo/http.py", line 310, in _handle_exception
>     raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
>   File "/data/build/odoo/odoo/tools/pycompat.py", line 14, in reraise
>     raise value
>   File "/data/build/odoo/odoo/http.py", line 669, in dispatch
>     result = self._call_function(**self.params)
>   File "/data/build/odoo/odoo/http.py", line 350, in _call_function
>     return checked_call(self.db, *args, **kwargs)
>   File "/data/build/odoo/odoo/service/model.py", line 94, in wrapper
>     return f(dbname, *args, **kwargs)
>   File "/data/build/odoo/odoo/http.py", line 339, in checked_call
>     result = self.endpoint(*a, **kw)
>   File "/data/build/odoo/odoo/http.py", line 915, in __call__
>     return self.method(*args, **kw)
>   File "/data/build/odoo/odoo/http.py", line 515, in response_wrap
>     response = f(*args, **kw)
>   File "/data/build/odoo/addons/website_sale_coupon_delivery/controllers/main.py", line 44, in cart_carrier_rate_shipment
>     return super(WebsiteSaleCouponDelivery, self).cart_carrier_rate_shipment(carrier_id, **kw)
>   File "/data/build/odoo/odoo/http.py", line 515, in response_wrap
>     response = f(*args, **kw)
>   File "/data/build/odoo/addons/website_sale_delivery/controllers/main.py", line 35, in cart_carrier_rate_shipment
>     assert int(carrier_id) in order._get_delivery_methods().ids, "unallowed carrier"
> AssertionError: unallowed carrier

</details>